kube-controllers/pkg/controllers/node/ipam.go

```diff
@@ -918,7 +917,7 @@ func (c *ipamController) allocationIsValid(a *allocation, preferCache bool) bool
 	}
 	if p == nil {
 		logc.Debug("Querying Kubernetes API for pod")
-		p, err = c.clientset.CoreV1().Pods(ns).Get(context.Background(), pod, metav1.GetOptions{})
+		p, err = c.podLister.Pods(ns).Get(pod)
```

Comment:
So, the way this code was originally intended to work is a bit confusing. Here's what I remember:

The controller has its own `c.podCache`, which is populated with pods as we get events from the pod informer. It's inefficient, because we're keeping two copies of each pod (one within the informer's cache and another in this cache) but it allowed us to prime this cache on misses ourselves (as is done on line 932).

The reason we have the clientset call to get the pod here instead of always using the pod lister as you have done here is because we were worried about scenarios where the informer might be behind actual cluster state, and we wanted a way to be 100% confident that the state of the pod was what we thought it was.

The intention was that this line would only be hit in rare situations, since most of the time podCache would be populated by the informer and we would have a cache hit, skipping this branch. But it seems that's not what you're experiencing, so I wonder why that is.

The only other time we would hit this is if `preferCache` is `false`, which is set [here](https://github.com/projectcalico/calico/blob/f413cf361aaceb84fa12696468d7cc08e72ca2b4/kube-controllers/pkg/controllers/node/ipam.go#L1056) only after we have made a decision to garbage collect an IP address, so we can do a final check that we're not acting on obsolete information about the pod. The assumption here again was that this would be called infrequently enough that a direct pod GET wouldn't have much cost.

All of this clunky code was a trade-off to increase our confidence, since the action of garbage collecting an IP address can be pretty problematic if the IP address itself is actually still in use somehow!

All of that said, I am not necessarily against the changes in this PR, but I need to think through them a little more and I think there are probably some other consequences. For example, I don't think `podCache` is serving any purpose any more with your changes so can probably be removed entirely. The other thing I need to consider is whether or not we still want that final escape hatch for `preferCache=false` from line 1056.

Reply:
>  I'm thinking how we can solve the issue I've outlined above while maintaining consistent state. Open to any suggestions.

How about go to informer indexer if `perferCache=true` and do standalone GET otherwise (on GC)? I need to test this but I think this should do it.

Reply:
> This causes thousands of GETs (one for each pod) in large clusters and informer falls far behind. I suspect the clientset also self-throttles itself and that affects the informer as well.

Ick! Yep, that definitely wasn't the original intention at least. We might be missing a signal somewhere to indicate that the informer has finished synchronizing? 

> How about go to informer indexer if perferCache=true and do standalone GET otherwise (on GC)? I need to test this but I think this should do it.

This sounds like a good approach to me :+1: